pkg/runtime/sequencer/cache.go

```diff
@@ -2,15 +2,151 @@ package sequencer
 
 import (
 	"context"
+	"errors"
 	"mosn.io/layotto/components/sequencer"
+	"mosn.io/pkg/log"
+	"mosn.io/pkg/utils"
+	"sync"
 )
 
+const defaultSize = 1000
+const defaultLimit = 700
+
+// DoubleBuffer is double segment id buffer.
+// There are two buffers in DoubleBuffer: InUseBuffer is in use, BackUpBuffer is a backup buffer.
+// Their default capacity is 1000. When the InUseBuffer usage exceeds 30%, the BackUpBuffer will be initialized.
+// When InUseBuffer is used up, swap them.
+type DoubleBuffer struct {
+	Key          string
+	Size         int
+	InUseBuffer  *Buffer
+	BackUpBuffer *Buffer
+	lock         sync.Mutex
+	Store        sequencer.Store
+}
+
+type Buffer struct {
+	from int64
+	to   int64
+}
+
+func NewDoubleBuffer(key string, store sequencer.Store) *DoubleBuffer {
+
+	d := &DoubleBuffer{
+		Key:   key,
+		Size:  defaultSize,
+		Store: store,
+	}
+
+	return d
+}
+
+//init double buffer
+func (d *DoubleBuffer) init() error {
+
+	buffer, err := d.getNewBuffer()
+	if err != nil {
+		return err
+	}
+
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	d.InUseBuffer = buffer
+
+	return nil
+}
+
+//getId next id
+func (d *DoubleBuffer) getId() (int64, error) {
+
+	if d.InUseBuffer == nil {
+		return 0, errors.New("[DoubleBuffer] Get error: InUseBuffer nil ")
+	}
+
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	next := d.InUseBuffer.from
+	d.InUseBuffer.from++
+
+	//check swap
+	if d.InUseBuffer.from > d.InUseBuffer.to {
+		d.swap()
+	}
+
+	//when InUseBuffer id more than half used, initialize BackUpBuffer.
+	//equal make sure only one thread enter
+	if d.InUseBuffer.to-d.InUseBuffer.from == defaultLimit {
+		utils.GoWithRecover(func() {
+
+			buffer, err := d.getNewBuffer()
+			if err != nil {
+				log.DefaultLogger.Errorf("[DoubleBuffer] [getNewBuffer] error: %v", err)
+			}
+			d.BackUpBuffer = buffer
```

Comment:
buffer may be nil

Reply:
@MoonShining  yes, I don't know what to do when an asynchronous error occurs？maybe try again？

Reply:
how about check BackUpBuffer exist before swap

Reply:
@MoonShining  ok, check buffer is nil, we get it again , if still nil  then return error